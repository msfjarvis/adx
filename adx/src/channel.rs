use crate::version::Version;
use clap::ValueEnum;
use clap::builder::PossibleValue;
use semver::Prerelease;
use std::cmp::{Eq, PartialEq, PartialOrd};
use std::convert::TryFrom;
use std::fmt::Debug;
use std::str::FromStr;
use thiserror::Error;

/// Release channels for androidx packages
/// Since we're deriving `PartialOrd` automatically, the order
/// of these fields is crucial. Sort by stability, not alphabetical
/// order.
#[derive(Debug, Eq, PartialEq, PartialOrd, Clone, Copy)]
pub(crate) enum Channel {
    Dev,
    Alpha,
    Beta,
    Rc,
    Stable,
}

impl ValueEnum for Channel {
    fn value_variants<'a>() -> &'a [Self] {
        &[Self::Dev, Self::Alpha, Self::Beta, Self::Rc, Self::Stable]
    }

    fn to_possible_value(&self) -> Option<PossibleValue> {
        match self {
            Channel::Dev => Some(PossibleValue::new("dev").alias("d")),
            Channel::Alpha => Some(PossibleValue::new("alpha").alias("a")),
            Channel::Beta => Some(PossibleValue::new("beta").alias("b")),
            Channel::Stable => Some(PossibleValue::new("stable").alias("s")),
            Channel::Rc => Some(PossibleValue::new("rc").alias("r")),
        }
    }
}

#[derive(Debug, Error)]
pub(crate) enum ChannelError {
    #[error("no match found")]
    NoMatchFound,
    #[error("failed to determine channel for {0}")]
    FailedToParseVersion(Version),
}

impl FromStr for Channel {
    type Err = ChannelError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "alpha" | "a" => Ok(Channel::Alpha),
            "beta" | "b" => Ok(Channel::Beta),
            "dev" | "d" => Ok(Channel::Dev),
            "rc" | "r" => Ok(Channel::Rc),
            "stable" | "s" => Ok(Channel::Stable),
            _ => Err(ChannelError::NoMatchFound),
        }
    }
}

impl TryFrom<&Version> for Channel {
    type Error = ChannelError;

    fn try_from(value: &Version) -> Result<Self, Self::Error> {
        match value {
            Version::SemVer(version) => {
                if version.pre == Prerelease::EMPTY {
                    Ok(Channel::Stable)
                } else {
                    let pre_str = version.pre.to_string();
                    if pre_str.starts_with("alpha") {
                        Ok(Channel::Alpha)
                    } else if pre_str.starts_with("beta") {
                        Ok(Channel::Beta)
                    } else if pre_str.starts_with("dev") {
                        Ok(Channel::Dev)
                    } else if pre_str.starts_with("rc") {
                        Ok(Channel::Rc)
                    } else {
                        Err(ChannelError::FailedToParseVersion(value.clone()))
                    }
                }
            }
            Version::CalVer(_) => Ok(Channel::Stable),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Channel;
    use crate::version::Version;
    use semver::{BuildMetadata, Prerelease, Version as Semver};
    use std::convert::TryFrom;

    #[test]
    fn alpha_version() {
        let v = Version::SemVer(Semver {
            major: 1,
            minor: 1,
            patch: 0,
            pre: Prerelease::new("alpha01").unwrap(),
            build: BuildMetadata::EMPTY,
        });
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Alpha, channel.unwrap());
    }

    #[test]
    fn beta_version() {
        let v = Version::SemVer(Semver {
            major: 1,
            minor: 1,
            patch: 0,
            pre: Prerelease::new("beta01").unwrap(),
            build: BuildMetadata::EMPTY,
        });
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Beta, channel.unwrap());
    }

    #[test]
    fn dev_version() {
        let v = Version::SemVer(Semver {
            major: 1,
            minor: 1,
            patch: 0,
            pre: Prerelease::new("dev01").unwrap(),
            build: BuildMetadata::EMPTY,
        });
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Dev, channel.unwrap());
    }

    #[test]
    fn rc_version() {
        let v = Version::SemVer(Semver {
            major: 1,
            minor: 1,
            patch: 0,
            pre: Prerelease::new("rc01").unwrap(),
            build: BuildMetadata::EMPTY,
        });
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Rc, channel.unwrap());
    }

    #[test]
    fn stable_version() {
        let v = Version::SemVer(Semver::new(1, 1, 1));
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Stable, channel.unwrap());
    }

    #[test]
    fn calver() {
        let v = Version::CalVer((2023, 4, 0));
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Stable, channel.unwrap());
    }

    #[test]
    fn cmp_channels() {
        assert!(Channel::Stable > Channel::Rc);
        assert!(Channel::Rc > Channel::Beta);
        assert!(Channel::Beta > Channel::Alpha);
        assert!(Channel::Alpha > Channel::Dev);
    }

    #[test]
    fn cmp_parsed_versions() {
        let stable = Channel::try_from(&Version::SemVer(Semver::parse("1.1.0").unwrap())).unwrap();
        let rc = Channel::try_from(&Version::SemVer(Semver::parse("1.1.0-rc01").unwrap())).unwrap();
        let alpha =
            Channel::try_from(&Version::SemVer(Semver::parse("1.1.0-alpha01").unwrap())).unwrap();
        let beta =
            Channel::try_from(&Version::SemVer(Semver::parse("1.1.0-beta01").unwrap())).unwrap();
        assert!(Channel::Stable >= stable);
        assert!(Channel::Stable >= rc);
        assert!(Channel::Stable >= alpha);
        assert!(Channel::Stable >= beta);
        assert!(Channel::Rc >= rc);
        assert!(Channel::Rc >= beta);
        assert!(Channel::Rc >= alpha);
        assert!(Channel::Beta >= beta);
        assert!(Channel::Beta >= alpha);
        assert!(Channel::Alpha >= alpha);
    }
}
