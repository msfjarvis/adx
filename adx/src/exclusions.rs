use crate::parse::get_packages;
use clap::{ValueEnum, builder::PossibleValue};
use std::fmt::Write;

#[derive(Debug, Eq, PartialEq, PartialOrd, Clone, Copy)]
pub(crate) enum PrintType {
    IncludeGroup,
    IncludeModule,
}

impl ValueEnum for PrintType {
    fn value_variants<'a>() -> &'a [Self] {
        &[Self::IncludeGroup, Self::IncludeModule]
    }

    fn to_possible_value(&self) -> Option<clap::builder::PossibleValue> {
        match self {
            PrintType::IncludeGroup => Some(PossibleValue::new("group")),
            PrintType::IncludeModule => Some(PossibleValue::new("module")),
        }
    }
}

pub(crate) async fn print_inclusions(print_type: PrintType) {
    let packages = get_packages().await;
    let Ok(packages) = packages else { return };
    let mut rules = String::new();
    match print_type {
        PrintType::IncludeGroup => {
            packages
                .into_iter()
                .map(|pkg| pkg.group_id)
                .collect::<std::collections::HashSet<_>>()
                .into_iter()
                .for_each(|group_id| {
                    let _ = writeln!(rules, "includeGroup(\"{group_id}\")");
                });
        }
        PrintType::IncludeModule => {
            for pkg in packages {
                let _ = writeln!(
                    rules,
                    "includeModule(\"{}\", \"{}\")",
                    pkg.group_id, pkg.artifact_id
                );
            }
        }
    }
    println!(
        "
    dependencyResolutionManagement {{
      repositories {{
        exclusiveContent {{
          forRepository {{ google() }}
          filter {{
            {rules}
          }}
        }}
      }}
    }}
    ",
    );
}
