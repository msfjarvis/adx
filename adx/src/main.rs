mod channel;
mod exclusions;
mod package;
mod parse;
mod version;

use crate::exclusions::print_inclusions;
use channel::Channel;
use clap::Parser;
use color_eyre::Result;
use exclusions::PrintType;

#[derive(Parser)]
#[command(author, version, about)]
pub(crate) struct Cli {
    /// search term to filter packages with
    #[arg(required = true)]
    pub(crate) search_term: String,
    /// the release channel to find packages from
    #[arg(value_enum, long, short, default_value_t = Channel::Alpha)]
    pub(crate) channel: Channel,
    /// search term to filter packages with
    #[arg(required = false, long, default_value_t = false)]
    pub(crate) print_includes: bool,
    /// stuff
    #[arg(value_enum, long, default_value_t = PrintType::IncludeGroup)]
    pub(crate) print_type: PrintType,
}

#[tokio::main]
async fn main() -> Result<()> {
    color_eyre::install()?;
    let cli = Cli::parse();
    if cli.print_includes {
        print_inclusions(cli.print_type).await;
        return Ok(());
    }
    let packages = parse::parse(&cli.search_term, cli.channel).await?;
    if packages.is_empty() {
        println!("No results found!");
    } else {
        for package in &packages {
            println!("{package}");
        }
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use super::Cli;
    use insta_cmd::{assert_cmd_snapshot, get_cargo_bin};
    use std::process::Command;

    #[test]
    fn cli_assert() {
        <Cli as clap::CommandFactory>::command().debug_assert();
    }

    #[test]
    fn cli_help() {
        assert_cmd_snapshot!(Command::new(get_cargo_bin("adx")).arg("--help"));
    }

    #[test]
    fn cli_search() {
        assert_cmd_snapshot!(Command::new(get_cargo_bin("adx")).arg("appcompat"));
    }

    #[test]
    fn cli_search_stable() {
        assert_cmd_snapshot!(
            Command::new(get_cargo_bin("adx"))
                .arg("--channel")
                .arg("stable")
                .arg("appcompat")
        );
    }

    #[test]
    fn cli_search_no_results() {
        assert_cmd_snapshot!(Command::new(get_cargo_bin("adx")).arg("qtc"));
    }
}
