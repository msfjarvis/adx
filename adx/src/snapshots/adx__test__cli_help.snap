---
source: adx/src/main.rs
info:
  program: adx
  args:
    - "--help"
---
success: true
exit_code: 0
----- stdout -----
Poll Google's Maven repository to fetch the latest versions of AndroidX packages

Usage: adx [OPTIONS] <SEARCH_TERM>

Arguments:
  <SEARCH_TERM>  search term to filter packages with

Options:
  -c, --channel <CHANNEL>        the release channel to find packages from [default: alpha] [possible values: dev, alpha, beta, rc, stable]
      --print-includes           search term to filter packages with
      --print-type <PRINT_TYPE>  stuff [default: group] [possible values: group, module]
  -h, --help                     Print help
  -V, --version                  Print version

----- stderr -----
